### PR TITLE
Add async flag to siac renter download

### LINKED ISSUE #2209: [feature] siac renter download allow to queue download async
Add a paramater to siac download to allow it to trigger the download to siad and then return without waiting for the file to be downloaded. There already is support to view status with siac renter downloads.

### PR DESCRIPTION
**Pull request by MSevey**
resolves #2209

### CODE REVIEW

**Comment by ChrisSchinnerl** [comment]
Does that still print the progress to stdout? If it is so fast that it is not noticeable we can probably leave it like that, but otherwise we can probably just avoid spawning the goroutine if `renterDownloadAsync` equals `true`.

**Comment by DavidVorick** [comment]
Wouldn't it be annoying for the progress to be printed to stdout if you were trying to do async download?

**Comment by DavidVorick** [comment]
Yeah I think sufficient to just skip the goroutine that prints out the progress

**Comment by tbenz9** [comment]
I agree it shouldn't print progress, but I think it might be nice to print a message when the file is done downloading.

**Comment by MSevey** [comment]
Ok so should it just be broken into an if statement then like this:
```
var err error
	if renterDownloadAsync {
		err = httpClient.RenterDownloadFullGet(path, destination, renterDownloadAsync)
	} else {
		done := make(chan struct{})
		go downloadprogress(done, path)

		err = httpClient.RenterDownloadFullGet(path, destination, renterDownloadAsync)
		close(done)
	}
```

**Comment by ChrisSchinnerl** [comment]
@MSevey you can do:
```
done := make(chan struct{})
if !renterDownloadAsync { 
  go downloadprogress(done, path)
}
err = httpClient.RenterDownloadFullGet(path, destination, renterDownloadAsync)
close(done)
```
It doesn't really hurt to not use a channel.

**Comment by MSevey** [comment]
Ok great

File: cmd/siac/rentercmd.go [hunk 1]
```diff
@@ -470,11 +470,14 @@ func renterfilesdeletecmd(path string) {
...
 	done := make(chan struct{})
-	go downloadprogress(done, path)
-
-	err := httpClient.RenterDownloadFullGet(path, destination, false)
+	if !renterDownloadAsync {
+		go downloadprogress(done, path)
+	}
+	err = httpClient.RenterDownloadFullGet(path, destination, renterDownloadAsync)
 	close(done)
+
```
Comment (ChrisSchinnerl) [inline cmd/siac/rentercmd.go]: Can you remove this newline? We want to make sure the line in which the `err` is assigned is close to the `if` that checks the error.

**Comment by lukechampine** [comment]
I don't see much advantage to this over `siac renter download >/dev/null &`, but I guess that isn't as easy to do on Windows.

It might make sense to invert the goroutines here:
```go
go func() {
    err = httpClient.RenterDownloadFullGet(path, destination, renterDownloadAsync)
    close(done)
}()
downloadprogress(done, path)
```
since arguably it's the download that's happening asynchronously, not the printing.
In general, when structuring goroutines, I think it's best for the longest-lived goroutine to be in the main body. `downloadprogress` by definition outlives `RenterDownloadFullGet`, so putting it on the main code path ensures that all other goroutines have exited before the function returns. Whereas the way it's currently written, the `downloadprogress` goroutine could outlive `renterfilesdownloadcmd`. (Not a huge deal in this case, since the program is going to exit immediately afterward, but you get the idea.)

**Pull request merged**